libcxx/include/__locale_dir/locale_base_api.h

```diff
@@ -35,71 +122,163 @@
+//
+using __locale_t = locale_t;
+
+#ifndef _LIBCPP_MSVCRT_LIKE
+inline _LIBCPP_HIDE_FROM_ABI __locale_t __uselocale(__locale_t __loc) { return uselocale(__loc); }
+#endif
+
+inline _LIBCPP_HIDE_FROM_ABI __locale_t __newlocale(int __category_mask, const char* __name, __locale_t __loc) {
+  return newlocale(__category_mask, __name, __loc);
+}
+
+inline _LIBCPP_HIDE_FROM_ABI void __freelocale(__locale_t __loc) { freelocale(__loc); }
+
+inline _LIBCPP_HIDE_FROM_ABI lconv* __localeconv(__locale_t& __loc) { return __libcpp_localeconv_l(__loc); }
+
+//
+// Strtonum functions
+//
+inline _LIBCPP_HIDE_FROM_ABI float __strtof(const char* __nptr, char** __endptr, __locale_t __loc) {
+  return strtof_l(__nptr, __endptr, __loc);
+}
+
+inline _LIBCPP_HIDE_FROM_ABI double __strtod(const char* __nptr, char** __endptr, __locale_t __loc) {
+  return strtod_l(__nptr, __endptr, __loc);
+}
+
+inline _LIBCPP_HIDE_FROM_ABI long double __strtold(const char* __nptr, char** __endptr, __locale_t __loc) {
+  return strtold_l(__nptr, __endptr, __loc);
+}
+
+inline _LIBCPP_HIDE_FROM_ABI long long __strtoll(const char* __nptr, char** __endptr, int __base, __locale_t __loc) {
+  return strtoll_l(__nptr, __endptr, __base, __loc);
+}
+
+inline _LIBCPP_HIDE_FROM_ABI unsigned long long
+__strtoull(const char* __nptr, char** __endptr, int __base, __locale_t __loc) {
+  return strtoull_l(__nptr, __endptr, __base, __loc);
+}
+
+//
+// Character manipulation functions
+//
+inline _LIBCPP_HIDE_FROM_ABI int __islower(int __ch, __locale_t __loc) { return islower_l(__ch, __loc); }
+inline _LIBCPP_HIDE_FROM_ABI int __isupper(int __ch, __locale_t __loc) { return isupper_l(__ch, __loc); }
+inline _LIBCPP_HIDE_FROM_ABI int __isdigit(int __ch, __locale_t __loc) { return isdigit_l(__ch, __loc); }
+inline _LIBCPP_HIDE_FROM_ABI int __isxdigit(int __ch, __locale_t __loc) { return isxdigit_l(__ch, __loc); }
+inline _LIBCPP_HIDE_FROM_ABI int __strcoll(const char* __s1, const char* __s2, __locale_t __loc) {
+  return strcoll_l(__s1, __s2, __loc);
+}
+inline _LIBCPP_HIDE_FROM_ABI size_t __strxfrm(char* __dest, const char* __src, size_t __n, __locale_t __loc) {
+  return strxfrm_l(__dest, __src, __n, __loc);
+}
+inline _LIBCPP_HIDE_FROM_ABI int __toupper(int __ch, __locale_t __loc) { return toupper_l(__ch, __loc); }
+inline _LIBCPP_HIDE_FROM_ABI int __tolower(int __ch, __locale_t __loc) { return tolower_l(__ch, __loc); }
+
+#if _LIBCPP_HAS_WIDE_CHARACTERS
+inline _LIBCPP_HIDE_FROM_ABI int __wcscoll(const wchar_t* __s1, const wchar_t* __s2, __locale_t __loc) {
+  return wcscoll_l(__s1, __s2, __loc);
+}
+inline _LIBCPP_HIDE_FROM_ABI size_t __wcsxfrm(wchar_t* __dest, const wchar_t* __src, size_t __n, __locale_t __loc) {
+  return wcsxfrm_l(__dest, __src, __n, __loc);
+}
+inline _LIBCPP_HIDE_FROM_ABI int __iswspace(wint_t __ch, __locale_t __loc) { return iswspace_l(__ch, __loc); }
+inline _LIBCPP_HIDE_FROM_ABI int __iswprint(wint_t __ch, __locale_t __loc) { return iswprint_l(__ch, __loc); }
```

Comment:
Before this commit landed, symbols like `iswspace_l` were only referenced if `_LIBCPP_WCTYPE_IS_MASK` was false.  But now it looks like the guard here is `_LIBCPP_HAS_WIDE_CHARACTERS`.  Is this intentional/expected?

If I made a patch to add `#if defined(_LIBCPP_HAS_WIDE_CHARACTERS) && !_LIBCPP_HAS_WIDE_CHARACTERS` to guard these implementations for `__iswspace` and friends -- would that be acceptable?

Reply:
Actually, I think it would be better if we tried to simplify this a bit. It looks like `_LIBCPP_WCTYPE_IS_MASK` and `iswupper_l` & friends were introduced in 0081892d335b22f2d3b0b92071f7e7fc9d0172ec. I don't fully understand what's the problem they were working around, but I *think* `iswctype_l` should do the same thing on all platforms that provide it (contrary to the commit message). So instead I'd be curious to try assuming that `iswctype_l` behaves like it does on FreeBSD and Apple, and see what happens in that case. That would allow us to remove many of these `__iswFOO` functions in one go, unconditionally. So, basically, I'd try reverting 0081892d335b22f2d3b0b92071f7e7fc9d0172ec and see what happens with the CI bots.

Alternatively, if that doesn't work out, we could potentially move from `iswctype_l` to `__iswFOO` on all platforms unconditionally.

TLDR I'd like to try simplifying this instead of adding even more complexity by tweaking the locale base API using configuration options, since it's already quite a mess.

Reply:
I actually just tried that out locally and I don't think that helps a lot, since we still reference all of these `__iswFOO` functions even after ripping out `_LIBCPP_WCTYPE_IS_MASK`.

I actually fail to see that we were not referencing `iswspace_l` outside of `_LIBCPP_WCTYPE_IS_MASK` before this patch. What's the actual issue you're running into and what's your configuration like?

Reply:
> What's the actual issue you're running into and what's your configuration like?

Well, when we cherry-pick this commit downstream, it fails to compile libc++ now because our C library doesn't have `iswctype_l` et al.

My assumptions regarding `_LIBCPP_WCTYPE_IS_MASK` were based on a brief inspection of the change here and maybe I misunderstood.  Let me try to reduce the problem and confirm whether indeed `_LIBCPP_WCTYPE_IS_MASK` is the critical factor or perhaps something else.

Reply:
IIUC we lack a platform-specific `__locale_dir/locale_base_api/foo.h` and that might be the source of our problem here.  I will explore creating one of those that accepts the locale from these functions and discards it, calling the corresponding non-`_l()` function for these missing ones.

Reply:
In general, this series of changes aims to make it a lot easier for platforms to define the exact base API they want. It may result in a bit more lines of code than reusing the BSD-like code path, but in the end I'd expect it to be more robust and simpler to understand if you define the API for your downstream platform from scratch.